File: Algorithm-Study/Python/Baekjoon Online Judge/BOJ_9012_괄호.py
# 접근방법
# 스택
# 여는 괄호와 닫는괄호가 쌍으로 갯수가 맞는지 아닌지 검사하기만 하면 된다.

T = int(input())
for tc in range(1,T+1):
    arr = list(input())


    stack = []
    result = "NO"
    for i in range(len(arr)):
        if stack == [] and arr[i] == ")":
            break
        elif arr[i] == "(":
            stack.append(arr[i])
        else:
            stack.pop(-1)
    else:
        if stack == []:
            result = "YES"

    print(result)

File: Algorithm-Study/JavaScript/Baekjoon Online Judge/BOJ_9012_괄호.js
const fs = require("fs");
let input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const T = input[0];
input = input.slice(1);

function solution(i) {
  let stack = [];
  for (inputVal of input[i]) {
    if (inputVal === "(") {
      stack.push(inputVal);
    } else if (stack.length !== 0 && inputVal === ")") {
      stack.pop();
    } else {
      return "NO";
    }
  }
  if (stack.length >= 1) {
    return "NO";
  } else {
    return "YES";
  }
}

for (let i = 0; i < input.length; i++) {
  console.log(solution(i));
}
